runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java

```diff
@@ -216,6 +219,34 @@ public CompositeBehavior enterCompositeTransform(TransformTreeNode node) {
       return CompositeBehavior.ENTER_TRANSFORM;
     }
 
+    private boolean shouldDiffer(TransformTreeNode node) {
```

Comment:
I don't see it that way.
The way I see it, according to the runner API a runner needs to support `GroupByKey`, and if it can (or needs), should optimize on `Combine.PerKey` (for example).
`Pipeline.PipelineVisitor.Defaults#enterCompositeTransform` is exactly about that - should the visitor enter the composite transform, or can the runner directly translate it, so why delegate to the translator (in the SparkRunner case) ?
This is exactly the case of "unwrapping" a composite transform.

This all feels awkward because it's very unstructured (sort of edge-case handling), but I raised this in the dev list.

Maybe the SDK/Runner API can flag a composite transform that should be "unwrapped" by the runner - it already holds all the information necessary, my code actually does it all from within the visitor.